### ISSUE TITLE
Performance issues in Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials

### ISSUE THREAD

**Issue by DLPerf** [TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials#192]
Hello! I've found two types of performance issues in your program: 

- `batch()` should be called before `map()`.
- `tf.Session` being defined repeatedly leads to incremental overhead.

You can make your program more efficient by fixing the above two problems. Here are [the tensorflow document](https://tensorflow.google.cn/guide/data_performance?hl=zh_cn#vectorized_mapping) and [the Stack Overflow post](https://stackoverflow.com/questions/48051647/tensorflow-how-to-perform-image-categorisation-on-multiple-images) to support this.

Below are detailed issues about **`batch()` should be called before `map()`**:

- tensorflow_dl_models/official/wide_deep/wide_deep.py: `dataset = dataset.batch(batch_size)`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/official/wide_deep/wide_deep.py#L194) should be called before `dataset = dataset.map(parse_csv, num_parallel_calls=5)`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/official/wide_deep/wide_deep.py#L189).
- tensorflow_dl_models/samples/outreach/blogs/blog_estimators_dataset.py: `dataset = dataset.batch(32)`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/samples/outreach/blogs/blog_estimators_dataset.py#L81) should be called before `dataset = (tf.data.TextLineDataset(file_path).skip(1).map(decode_csv))`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/samples/outreach/blogs/blog_estimators_dataset.py#L74).
- tensorflow_dl_models/samples/outreach/blogs/blog_custom_estimators.py: `.batch(32)`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/samples/outreach/blogs/blog_custom_estimators.py#L77) should be called before `.map(decode_csv, num_parallel_calls=4)`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/samples/outreach/blogs/blog_custom_estimators.py#L73).
- tensorflow_dl_models/samples/core/get_started/iris_data.py: `dataset = dataset.shuffle(1000).repeat().batch(batch_size)`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/samples/core/get_started/iris_data.py#L90) should be called before `dataset = dataset.map(_parse_line)`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/samples/core/get_started/iris_data.py#L87).

Besides, you need to check the function called in `map()`(e.g., `_parse_line` called in `dataset = dataset.map(_parse_line)`) whether to be affected or not to make the changed code work properly. For example, if `_parse_line` needs data with shape (x, y, z) as its input before fix, it would require data with shape (batch_size, x, y, z).

Below are detailed issues about **`tf.Session` being defined repeatedly**:

- tensorflow_dl_models/tutorials/image/cifar10/cifar10_eval.py: `with tf.Session() as sess:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/tutorials/image/cifar10/cifar10_eval.py#L71) is defined in function `eval_once`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/tutorials/image/cifar10/cifar10_eval.py#L62) which is repeatedly called in a loop `while True:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/tutorials/image/cifar10/cifar10_eval.py#L142).
- tensorflow_dl_models/research/object_detection/eval_util.py: `sess = tf.Session(master, graph=tf.get_default_graph())`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/object_detection/eval_util.py#L231) is defined in function `_run_checkpoint_once`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/object_detection/eval_util.py#L171) which is repeatedly called in a loop `while True:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/object_detection/eval_util.py#L375).
- tensorflow_dl_models/research/im2txt/im2txt/evaluate.py: `with tf.Session() as sess:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/im2txt/im2txt/evaluate.py#L122) is defined in function `run_once`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/im2txt/im2txt/evaluate.py#L107) which is repeatedly called in a loop `while True:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/im2txt/im2txt/evaluate.py#L184).
- tensorflow_dl_models/research/capsules/experiment.py: `session = tf.Session(config=tf.ConfigProto(allow_soft_placement=True))`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/capsules/experiment.py#L269) is defined in function `run_experiment`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/capsules/experiment.py#L243) which is repeatedly called in a loop `while paused < 360:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/capsules/experiment.py#L400).
- tensorflow_dl_models/research/street/python/vgsl_model.py: `sess = tf.Session('')`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/street/python/vgsl_model.py#L155) is defined in a loop `while True:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/street/python/vgsl_model.py#L154).
- tensorflow_dl_models/research/skip_thoughts/skip_thoughts/track_perplexity.py: `with tf.Session() as sess:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/skip_thoughts/skip_thoughts/track_perplexity.py#L122) is defined in function `run_once`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/skip_thoughts/skip_thoughts/track_perplexity.py#L105) which is repeatedly called in a loop `while True:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/skip_thoughts/skip_thoughts/track_perplexity.py#L194).
- tensorflow_dl_models/research/inception/inception/inception_eval.py: `with tf.Session() as sess:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/inception/inception/inception_eval.py#L65) is defined in function `_eval_once`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/inception/inception/inception_eval.py#L55) which is repeatedly called in a loop `while True:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/inception/inception/inception_eval.py#L168).
- tensorflow_dl_models/research/slim/datasets/download_and_convert_cifar10.py: `with tf.Session('') as sess:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/slim/datasets/download_and_convert_cifar10.py#L91) is defined in function `_add_to_tfrecord`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/slim/datasets/download_and_convert_cifar10.py#L64) which is repeatedly called in a loop `for i in range(_NUM_TRAIN_FILES):`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/tensorflow_dl_models/research/slim/datasets/download_and_convert_cifar10.py#L184).
- deep-learning/GANs and Variational Autoencoders/BigGAN-PyTorch/scripts/tfhub/converter.py: `sess = tf.Session()`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/deep-learning/GANs%20and%20Variational%20Autoencoders/BigGAN-PyTorch/scripts/tfhub/converter.py#L67) is defined in function `dump_tfhub_to_hdf5`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/deep-learning/GANs%20and%20Variational%20Autoencoders/BigGAN-PyTorch/scripts/tfhub/converter.py#L47) and `dump_tfhub_to_hdf5` is called in function `convert_biggan`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/deep-learning/GANs%20and%20Variational%20Autoencoders/BigGAN-PyTorch/scripts/tfhub/converter.py#L315) which is repeatedly called in a loop `for res in RESOLUTIONS:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/deep-learning/GANs%20and%20Variational%20Autoencoders/BigGAN-PyTorch/scripts/tfhub/converter.py#L396).
- deep-learning/udacity-deeplearning/weight-initialization/helper.py: `with tf.Session() as session:`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/deep-learning/udacity-deeplearning/weight-initialization/helper.py#L54) is defined in function `_get_loss_acc`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/deep-learning/udacity-deeplearning/weight-initialization/helper.py#L18) which is repeatedly called in a loop `for i, (weights, label) in enumerate(weight_init_list):`[(here)](https://github.com/TarrySingh/Artificial-Intelligence-Deep-Learning-Machine-Learning-Tutorials/blob/5a815b1429c9b3be3c4e192239488c141deeb00f/deep-learning/udacity-deeplearning/weight-initialization/helper.py#L98).

If you define `tf.Session` out of the loop and pass `tf.Session` as a parameter to the loop, your program would be much more efficient.

Looking forward to your reply. Btw, I am very glad to create a PR to fix it if you are too busy.

**Comment by DLPerf** [comment]
Hello, I'm looking forward to your reply~
